### PR TITLE
PXC-702: GCache encryption

### PR DESCRIPTION
**Pull request by kamil-holubicki**
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Server part implementation.
1. Needed changes around keyring component and Galera initialization
2. Master key rotation

### CODE REVIEW

File: plugin/keyring/common/keys_container.cc [hunk 1]
```diff
@@ -183,8 +206,11 @@ bool Keys_container::load_keys_from_keyring_storage() {
       if (serialized_keys->get_next_key(&key_loaded) || key_loaded == nullptr ||
           key_loaded->is_key_valid() == false ||
           store_key_in_hash(key_loaded)) {
-        was_error = true;
-        delete key_loaded;
+#ifdef WITH_WSREP
+        if (forceKeysFetch) {
```
Comment (dutow) [inline plugin/keyring/common/keys_container.cc]: Shouldn't `else` keep existing behavior?

File: plugin/keyring/common/keys_container.cc [hunk 2]
```diff
@@ -88,6 +92,12 @@ bool Keys_container::store_key_in_hash(IKey *key) {
 }
 
 bool Keys_container::store_key(IKey *key) {
+#ifdef WITH_WSREP
+  if (forceKeysFetch) {
+    load_keys_from_keyring_storage();
```
Comment (dutow) [inline plugin/keyring/common/keys_container.cc]: These hacks ignore the return value of load_keys_from_keyring_storage, which might fail, especially with a remote keyring.

**Review by dutow — COMMENTED** [review]
This seems too much of a hack to me.

Can't we explicitly reload the keyring cache after SST instead?

File: plugin/keyring/common/keys_container.cc (hunk 1 above)
Comment (kamil-holubicki) [reply plugin/keyring/common/keys_container.cc]: The part related to this hack has been completely reworked. Now Keyring plugin is reloaded after SST.

File: plugin/keyring/common/keys_container.cc (hunk 2 above)
Comment (kamil-holubicki) [reply plugin/keyring/common/keys_container.cc]: The part related to this hack has been completely reworked. Now Keyring plugin is reloaded after SST.

**Pull request closed without merging**